src/OrchardCore.Modules/OrchardCore.Users/Localization/Providers/UserLocalizationRequestCultureProvider.cs

```diff
@@ -0,0 +1,51 @@
+using System;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Localization;
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Localization;
+
+namespace OrchardCore.Users.Localization.Providers;
+
+public class UserLocalizationRequestCultureProvider : RequestCultureProvider
+{
+    public override async Task<ProviderCultureResult> DetermineProviderCultureResult(HttpContext httpContext)
+    {
+        ArgumentNullException.ThrowIfNull(httpContext);
+
+        if (httpContext?.User?.Identity?.IsAuthenticated == false)
+        {
+            return await NullProviderCultureResult;
+        }
+
+        string userCulture = null;
+        var claimCulture = httpContext.User.FindFirstValue("culture");
+
+        if (!String.IsNullOrWhiteSpace(claimCulture))
+        {
+            userCulture = claimCulture;
+        }
+
+        if (userCulture == null)
+        {
+            // No values specified for either so no match
+            return await NullProviderCultureResult;
+        }
+
+        var localizationService = httpContext.RequestServices.GetService<ILocalizationService>();
+
+        var supportedCulture = await localizationService.GetSupportedCulturesAsync();
+
+        // We verify that the userCulture is still a supportedCulture
+        if (!supportedCulture.Contains(userCulture))
+        {
+            return await NullProviderCultureResult;
```

Comment:
Yes finally I agree with @hishamco about returning `NullProviderCultureResult`, the github searching in the aspnetcore repo no longer works for me ;)

Reply:
I just worked on the repo recently

https://github.com/dotnet/aspnetcore/blob/888c71f7f2269b6e67a24924cef66228857236f3/src/Middleware/Localization/src/RequestCultureProvider.cs#L18

Reply:
Yes but finally the aspnet middleware also does a null check, so that's fine and finally better for perf to return null ;)

Reply:
> Yes but finally the aspnet middleware also does a null check, so that's fine and finally better for perf to return null ;)

Seems they added `ProviderCultureResult?` lately, but we might return `null` if returns `ProviderCultureResult` from a method, then return to the caller in `DetermineProviderCultureResult`

Reply:
Yes, as said the github search work everywhere unless in the aspnetcore repo for me ;) yes I saw that it's like an helper to use a `Task.FromResult()` when in a method you don't use any `await`.

    NullProviderCultureResult = Task.FromResult(default(ProviderCultureResult));

But when your method already use an `await`, in place of awaiting a `Task.FromResult()`, better to just return `default(ProviderCultureResult)`, so here the type being implied, my suggestion is.

    return default;


 Which returns null because `ProviderCultureResult` is a class (not a value type) so `return null;` was also okay.